File: implementations/android/client/app/src/main/java/com/theshuai/specus/android/PeerServiceResourceLimiter.java
package com.theshuai.specus.android;

import java.net.InetAddress;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.Semaphore;
import java.util.concurrent.atomic.AtomicBoolean;

final class PeerServiceResourceLimiter {
    static final int MAX_TCP_GLOBAL = 256;
    static final int MAX_TCP_PER_SERVICE = 64;
    static final int MAX_TCP_PER_SOURCE = 8;
    static final int MAX_UDP_GLOBAL = 256;
    static final int MAX_UDP_PER_SERVICE = 64;
    static final int MAX_UDP_PER_SOURCE = 8;
    static final int CONNECT_TIMEOUT_MILLIS = 3_000;
    static final int IDLE_TIMEOUT_MILLIS = 60_000;

    private static final Object LOCK = new Object();
    private static final Semaphore TCP_GLOBAL = new Semaphore(MAX_TCP_GLOBAL);
    private static final Semaphore UDP_GLOBAL = new Semaphore(MAX_UDP_GLOBAL);
    private static final Map<InetAddress, Integer> TCP_SOURCES = new HashMap<>();
    private static final Map<InetAddress, Integer> UDP_SOURCES = new HashMap<>();

    private PeerServiceResourceLimiter() {
    }

    static Lease tryAcquireTcp(InetAddress source) {
        return acquire(source, TCP_GLOBAL, TCP_SOURCES, MAX_TCP_PER_SOURCE);
    }

    static Lease tryAcquireUdp(InetAddress source) {
        return acquire(source, UDP_GLOBAL, UDP_SOURCES, MAX_UDP_PER_SOURCE);
    }

    private static Lease acquire(InetAddress source, Semaphore global, Map<InetAddress, Integer> sources, int limit) {
        if (source == null || !global.tryAcquire()) {
            return null;
        }
        synchronized (LOCK) {
            int active = sources.getOrDefault(source, 0);
            if (active >= limit) {
                global.release();
                return null;
            }
            sources.put(source, active + 1);
        }
        return new Lease(source, global, sources);
    }

    static final class Lease implements AutoCloseable {
        private final InetAddress source;
        private final Semaphore global;
        private final Map<InetAddress, Integer> sources;
        private final AtomicBoolean closed = new AtomicBoolean();

        private Lease(InetAddress source, Semaphore global, Map<InetAddress, Integer> sources) {
            this.source = source;
            this.global = global;
            this.sources = sources;
        }

        @Override
        public void close() {
            if (!closed.compareAndSet(false, true)) {
                return;
            }
            synchronized (LOCK) {
                int remaining = sources.getOrDefault(source, 1) - 1;
                if (remaining <= 0) {
                    sources.remove(source);
                } else {
                    sources.put(source, remaining);
                }
            }
            global.release();
        }
    }
}

File: implementations/csharp/client/src/Specus.Client/PeerMesh/PeerServiceResourceLimiter.cs
using System.Net;

namespace Specus.Client.PeerMesh;

internal static class PeerServiceResourceLimiter
{
    internal const int MaxTcpGlobal = 256;
    internal const int MaxTcpPerService = 64;
    internal const int MaxTcpPerSource = 8;
    internal const int MaxUdpGlobal = 256;
    internal const int MaxUdpPerService = 64;
    internal const int MaxUdpPerSource = 8;
    internal static readonly TimeSpan ConnectTimeout = TimeSpan.FromSeconds(3);
    internal static readonly TimeSpan IdleTimeout = TimeSpan.FromSeconds(60);

    private static readonly object Gate = new();
    private static readonly SemaphoreSlim TcpGlobal = new(MaxTcpGlobal, MaxTcpGlobal);
    private static readonly SemaphoreSlim UdpGlobal = new(MaxUdpGlobal, MaxUdpGlobal);
    private static readonly Dictionary<IPAddress, int> TcpSources = [];
    private static readonly Dictionary<IPAddress, int> UdpSources = [];

    internal static Lease? TryAcquireTcp(IPAddress source) =>
        TryAcquire(source, TcpGlobal, TcpSources, MaxTcpPerSource);

    internal static Lease? TryAcquireUdp(IPAddress source) =>
        TryAcquire(source, UdpGlobal, UdpSources, MaxUdpPerSource);

    private static Lease? TryAcquire(IPAddress source, SemaphoreSlim global,
        Dictionary<IPAddress, int> sources, int sourceLimit)
    {
        source = Normalize(source);
        if (!global.Wait(0))
        {
            return null;
        }
        lock (Gate)
        {
            sources.TryGetValue(source, out var active);
            if (active >= sourceLimit)
            {
                global.Release();
                return null;
            }
            sources[source] = active + 1;
        }
        return new Lease(source, global, sources);
    }

    private static IPAddress Normalize(IPAddress address) =>
        address.IsIPv4MappedToIPv6 ? address.MapToIPv4() : address;

    internal sealed class Lease(IPAddress source, SemaphoreSlim global, Dictionary<IPAddress, int> sources)
        : IDisposable
    {
        private int _disposed;

        public void Dispose()
        {
            if (Interlocked.Exchange(ref _disposed, 1) != 0)
            {
                return;
            }
            lock (Gate)
            {
                var remaining = sources.GetValueOrDefault(source) - 1;
                if (remaining <= 0)
                {
                    sources.Remove(source);
                }
                else
                {
                    sources[source] = remaining;
                }
            }
            global.Release();
        }
    }
}
